File: PySCT/Options.py

class Options:

    # Constructors & Destructor
    def __init__( self, id = None, truth = None, val = None ):
        
        self._opt: str = id if id is not None else "NULL"

        self._stats: bool = truth if truth is not None else False

        self._value: int = val if val is not None else -1

    def __del__( self ):

        self._opt = ""

        self._stats = False

        self._value = -1

    # Setters
    def set_opt( self, id ):

        self._opt = id

    def set_status( self, truth ):

        self._stats = truth

    def set_value( self, val ):

        self._value = val

    # Getters
    def get_opt( self ) ->  str:

        return self._opt
    
    def get_status( self ) -> bool:

        return self._stats

    def get_value( self ) -> int:

        return self._value

    # Operators

    # Helpers
    def __str__( self ):

        return "( " + str( self._opt ) + ", " + str( self._value ) + " )"

    def __lt__( self, right ):

        if self._value < right._value:

            return True
        
        else:

            return False


# Testing methods below

def test1():
    
    x = Options( "a", False, 5 )

    print( Options( "b", True, 56 ) )

    y = x

    print( y )

    x.set_opt( "D" )
    x.set_status( True )
    x.set_value( 84 )

    print( x )

#test1()

#x = Options( )

#print( x )

File: options.cpp
#include "/Projetos/SocialChoice_VS/Include/options.hpp"

/// Constructors & Destructor

// Default constructor. Initializes private properties to its default values
Options::Options( ){

    opt = "NULL";

    status = false;

    value = -1;
}

Options::Options( std::string name ){

	opt = name;

	status = false;

	value = -1;
}

// Copy constructor. Initializes private properties based on the arguments passed by
Options::Options( const Options& copy ){

	opt = copy.opt;
    status = copy.status;
	value = copy.value;
}

Options::Options( Options&& copy ){

	opt = std::move( copy.opt );

	status = std::move( copy.status );

	value = std::move( copy.value );

	copy.clear( );
}

Options::~Options( ){ clear( ); }

/// Setters

// Sets alternatives according to the argument passed through
void Options::set_opt( std::string newalt ){ opt = newalt; }

void Options::set_status( bool value  ){ status = value; }

// Sets values according to value passed as parameter
void Options::set_value( int val ){ value = val; }

/// Operators

// Overloaded assignment operator
Options& Options::operator=( const Options& copy ){

	opt = copy.opt;

	status = copy.status;

	value = copy.value;

	return *this;
}

Options& Options::operator=( Options&& copy ){

	opt = std::move( copy.opt );

	status = std::move( copy.status );

	value = std::move( copy.value );

	copy.clear( );

	return *this;
}

// Overloaded increment operator. TEST THIS
Options& Options::operator++( ){

    //value = value + 1;
    ++value;

    return *this;
}
Options Options::operator++( int value ){

	Options tmp( *this );

	operator++( );

	return tmp;
}

/// Helpers

void swap( Options& left, Options& right ){

	using std::swap;

	swap( left.opt, right.opt );
	swap( left.status, right.status );
	swap( left.value, right.value );
}

// Resets an instance to its default state
void Options::clear( ){

	opt.clear( );

	status = false;

	value = -1;
}

// Overloaded assign-increment operator. TEST THIS
Options& Options::operator+=( const Options& val ){ 
	
	value += val.value;

	return *this;
}
Options& Options::operator+=( const int val ){ 
	
	value += val;

	return *this;
}

/// Non-member helpers

// Prints option
std::ostream& operator<<( std::ostream& os, Options& opt ){

    os << "[ " << opt.get_opt( ) << ", " << opt.get_value( ) << " ]" ;

    return os;
}
